[Week2] Spring Web으로 ToDo REST API 만들기

JIY0UNG opened the pull request:
안녕하세요!!
Task 전체 조회, 상세 조회, 생성, 제목 수정, 삭제 작업을 처리하는 TaskController를 만들었습니다. 리뷰 부탁드립니다😊

그리고 1주 차 때 다른 분들의 PR을 보다가 [윤석님께서 커밋 메시지 작성법에 대해 말씀해 주신 것](https://github.com/CodeSoom/spring-week1-assignment-1/pull/95#issuecomment-1118089172)을 참고하여 제 커밋 메시지에도 한번 적용해 보았어요!

JIY0UNG committed:
Spring Developer Tools를 사용하기 위한 Gradle 설정을 추가하라

좀 더 편리하게 개발하기 위해, 빌드만 하면 서버를 자동으로 재시작 해주는 도구인 Spring Developer Tools를 사용할 수 있도록 Gradle 설정을 추가하였습니다.

JIY0UNG committed:
불필요한 getGreeting 메서드와 HelloController 클래스를 삭제하라

App 클래스의 getGreeting 메서드와 HelloController 클래스 모두 사용하지 않는 것들이기 때문에 삭제하였습니다.

JIY0UNG committed:
TaskController 클래스를 생성하고, Task 목록을 얻어오는 코드를 작성하라

JIY0UNG committed:
TaskController에 Task를 생성하는 코드를 작성하라

- Task 생성 시 title이 비어있으면 응답 코드 400을 반환하도록 하였습니다.
- Task가 올바르게 생성된 후에는 응답 코드 201을 반환하도록 하였습니다.

JIY0UNG committed:
TaskController에 id로 특정 Task에 대한 상세 조회를 하는 코드를 작성하라

Task 조회에 실패한 경우 응답 코드 404를 반환하도록 하였습니다.

JIY0UNG committed:
TaskController에 Task의 title을 수정하는 코드를 작성하라

- 새로운 title을 입력하지 않았을 경우 응답 코드 400을 반환하도록 하였습니다.
- 수정하려는 Task 조회에 실패한 경우 응답 코드 404를 반환하도록 하였습니다.

JIY0UNG committed:
TaskController에 Task를 삭제하는 코드를 작성하라

삭제하려는 Task 조회에 실패한 경우 응답 코드 404를 반환하도록 하였습니다.

JIY0UNG committed:
Task 삭제에 성공하면 응답 코드 204를 반환하도록 하라

기존에는 삭제 성공 시 별도로 응답 코드를 설정해 주지 않아서 200 코드가 반환 됐었는데, PR build 로그에서 기대했던 응답 코드는 204인데 실제 응답 코드는 200이어서 API 테스트에 실패했다는 것을 발견했습니다.
그래서 성공 시 반환할 응답 코드를  204로 설정하도록 하였습니다.

bloomspes commented:
안녕하세요 지영님, 만나서 반갑습니다. 커밋 메시지를 작성할 때도 관심을 많이 기울이고 계시네요. 👍🏻
2주 차 인데, 적응을 잘 하고 계신 것 같아요. 이번 주도 함께 화이팅 해요. 🤙🏻

JIY0UNG committed:
Task 삭제 요청 uri에 id가 존재하지 않을 경우 응답 코드 404를 반환하도록 하라

기존에는 id를 무조건 입력받는다는 가정 하에 코드를 작성했기 때문에, id가 입력되지 않았을 경우에 대한 처리를 별도로 하지 않았었습니다.
PR build 로그에서 이런 경우에 기대했던 응답 코드는 404인데 실제 응답 코드는 204여서 API 테스트에 실패했다는 것을 발견했고, 그래서 id가 존재하지 않을 경우 응답 코드를 404로 설정하도록 하였습니다.

bloomspes commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,20 @@
+package com.codesoom.assignment.controllers;
+
+import com.codesoom.assignment.models.Task;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@RestController
+@RequestMapping("/tasks")
+public class TaskController {
+    private List<Task> tasks = new ArrayList<>();
+
+    @GetMapping
+    public List<Task> getTaskList() {
+        return tasks;
+    }
+}
```
이번 주 부터 많은 어노테이션을 공부하고 계시네요! 👍🏻 사용하신 모든 어노테이션에 대해 짧게 정리해보세요.
`@RequestMapping` 어노테이션은 메소드 레벨에서 사용하면 어떤 결과가 나올지 직접 확인해보세요.

JIY0UNG replied:
현재 컨트롤러에서 사용중인 어노테이션들에 대해 정리해봤습니다. 틀린 부분이 있으면 말씀해주시면 감사하겠습니다!!

`@RestController`
- 해당 클래스가 REST API를 구현한 컨트롤러임을 나타낸다.
- `@Controller`와는 다르게 RequestBody를 객체로 받기 위해서 메서드에 `@ResponseBody`를 붙여줄 필요가 없다.
- 
`@RequestMapping`
- 특정 URL을 클래스나 핸들러 메서드에 매핑해주는 역할을 한다.
- 메서드 레벨에서 path 속성을 통해 각자 매핑할 URL을 설정할 수 있다. (보통 URL에서 공통되는 부분을 클래스 레벨로 빼서 정의한다.)
- 같은 URL에 대해 HTTP 메서드별로 다르게 매핑 하고 싶을 때는 method 속성으로 각각 HTTP 메서드를 설정해 줘야 매핑이 이루어진다. (안 그러면 이미 bean으로 등록돼 있다면서 BeanCreationException 에러가 발생한다.)

`@GetMapping`, `@PostMapping`, `@PutMapping`, `@PatchMapping` , `@DeleteMapping`
- 각각 HTTP GET, POST, PUT, PATCH, DELETE 요청을 처리한다. (`@RequestMapping`의 속성으로 method를 설정해준 것과 같다.)
- 주의) PUT은 데이터 전체를 수정하고, PATCH는 일부 영역만 수정한다.

bloomspes replied:
컨트롤러에 대해 공부하셨군요. 👍🏻 정리해주신 부분에서 중요한 부분을 찾을 수 있었어요.

bloomspes replied:
> `@Controller`와는 다르게 RequestBody를 객체로 받기 위해서 메서드에 `@ResponseBody`를 붙여줄 필요가 없다.

그 이유는, `@RestController`이 `@Controller` 와 `@RequestBody`의 주요 기능을 묶어 놓았기 때문이에요.
아래의 Spring Docs를 보시면 정리가 잘 되실거예요. 
새로운 인터페이스와 메소드를 공부하실 때마다 Docs를 확인하시는 습관을 들이신다면 정말 많은 도움이 될 거예요.

- [RestController 에 대해 읽기](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/web/bind/annotation/RestController.html)

그러면 `@RequestMapping`에 대해서도 default로 어떤 인자를 받을 수 있는지 알게 되고, 
지영님이 정리를 잘 해주신 것처럼 메서드 레벨에서 path 속성을 통해 각자 매핑할 URL을 설정할 수 있게 되죠.

정리해주신 부분을 읽으면서 인상깊은 부분을 발견했는데요,
> 안 그러면 이미 bean으로 등록돼 있다면서 BeanCreationException 에러가 발생한다.

라고 덧 붙여서 남겨주셨어요. 공부하실 때 어떤 결과를 리턴하는 지 확인하고 정리하는 부분이 참 좋습니다.

또 한가지, docs를 읽다보면 숨겨진 사실을 알게 됩니다.
>  `@GetMapping` is a composed annotation that acts as a shortcut for `@RequestMapping(method = RequestMethod.GET)`.

`PUT` 과 `PATCH`에 대해서도 꼼꼼하게 정리를 해주셨어요. 👍🏻

bloomspes commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -18,6 +18,18 @@ public List<Task> getTaskList() {
         return tasks;
     }
 
+    @GetMapping("{id}")
+    public Object getTaskItem(@PathVariable("id") Long id, HttpServletResponse response) {
+        Task task = findTask(id);
+
+        if (task == null) {
+            response.setStatus(404);
+            return "";
+        }
+
+        return task;
+    }
```
id로 task를 조회하고, task가 비어있는 경우, response 객체로 404 응답을 가져오고, 빈 문자열을 리턴해주고 있어요.

컨트롤러에선 직접 일을 수행하지 않고, 서비스 혹은 비즈니스 로직에 작업을 위임하도록 합니다.
컨트롤러는 왜 이렇게 계층을 나눠서 작업해야 하는 걸까요? 
이에 대해 자세히 정리하고 싶다면, 아래의 종립님의 블로그를 참조하시면 도움이 될 거예요.

우선은, `Task` 객체가 하는 일을 TaskController 에서 분리를 시도해보세요. 

- [컨트롤러 패턴 (Controller Pattern)](https://johngrib.github.io/wiki/pattern/controller/)

JIY0UNG replied:
글을 읽어보니 지금 제 컨트롤러는 '팽창된 컨트롤러'라고 할 수 있겠네요.
기존에 컨트롤러가 하던 일들을 Service 클래스로 분리하도록 하겠습니다!

JIY0UNG replied:
> 크레이그 라만은 다음과 같은 해결책을 제안한다.
> 1. 더 많은 컨트롤러를 추가한다.
> 2. 각 시스템 오퍼레이션에 대한 책임 수행을 다른 객체에게 위임하도록 controller를 설계한다.

글에서 말한 해결책 중 1번을 선택할 경우에는,
결국 또 다른 컨트롤러에서 일을 수행하게 되므로 '컨트롤러가 직접 일을 수행하지 않게' 만드는 것에 부합하지 않는 것 아닌가?라는 생각이 들었는데, 혹시 제가 잘못 이해한 걸까요??

bloomspes replied:
피드백을 잘 수용하시고 따라오고 계시네요, 열심히 공부하시는 모습이 참 멋집니다. :) 

하나의 컨트롤러에서 `모든` 요청을 받게 되면, 많은 요청이 컨트롤러 클래스 하나에만 몰리게 되죠. 
이런 경우도 위에 열거된 팽창된 컨트롤러에 부합하는 예가 될 거예요.

예를 들어서, 사용자 정보에 대한 요청과 프로덕트에 대한 요청은 하나의 컨트롤러에서 처리하기에는 많은 책임을 안게 되겠죠.
이런 경우라면 사용자 정보를 얻는 요청과 프로덕트 정보를 얻는 요청을 분리하는게 좋을거에요. 

그런 맥락에서의 더 많은 컨트롤러를 추가한다. 라고 이해하시면 도움이 될 거예요.

JIY0UNG replied:
아아 지금은 Task에 대한 요청만 받고 있기 때문에 1개의 컨트롤러만 쓰고 있지만,
나중에 Task 외의 요청도 받게 된다면 2개 이상의 컨트롤러를 써야 한다는 말과 같군요!

bloomspes commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -44,6 +44,18 @@ public Object createTask(@RequestBody Task task, HttpServletResponse response) {
         }
     }
 
+    @PutMapping("/{id}")
+    public Object updateTaskByPutMethod(@PathVariable("id") Long id,
+                                        @RequestBody Task source, HttpServletResponse response) {
+        return handleUpdate(id, source, response);
+    }
+
+    @PatchMapping("/{id}")
+    public Object updateTaskByPatchMethod(@PathVariable("id") Long id,
+                                          @RequestBody Task source, HttpServletResponse response) {
+        return handleUpdate(id, source, response);
+    }
+
```
여기도 아래의 코멘트를 참조해서 책임을 분리할 수 있을거예요. 
https://github.com/CodeSoom/spring-week2-assignment-1/pull/82#discussion_r869275209

메소드 명은 '~ 메소드에 의해' 라고 붙여주시는 것 보다, 기능을 표현하는 단어를 찾도록 같이 연습해봐요.
메소드의 역할을 명확히 드러낼 수 있어요.

우리가 사용하는 메소드는 이 후에 리팩토링을 거쳐서 사용하지 않을 수도 있고, 구조가 바뀌면 
메소드 이름도 바꿔야 하는 상황이 올 수 있겠죠?

`updateTaskByPutMethod`는 `update`로, `updateTaskByPatchMethod` 는 `patch` 로 변경해도 의미를 충분히 전달 할 수 있을 거예요.

JIY0UNG replied:
훨씬 간단하면서도 메서드의 기능을 잘 나타낼 수 있는 이름이네요! 수정하도록 하겠습니다. 다른 메서드들도 기능을 잘 나타내는 더 좋은 이름이 있을지 생각해 봐야겠어요!

JIY0UNG replied:
> 여기도 아래의 코멘트를 참조해서 책임을 분리할 수 있을거예요.

여기서 책임을 더 분리하려면, 컨트롤러 내의 handleUpdate 메서드를 호출하던 것을 Service 클래스 내의 handleUpdate 메서드를 호출하도록 바꾸면 될까요??
이미 한 줄짜리 코드라 어떻게 더 분리해야 할지 좀 감이 안 오네요ㅜ

bloomspes replied:
맞습니다. 컨트롤러 안에 있는 handleUpdate 메소드가 Service 클래스 안에서 호출되어야 할 거에요.
정확히는 한 줄 짜리 코드여서는 아니고, `handleUpdate` 메소드가 담고있는 정보가 많아서 분리가 어렵게 느껴지셨을거예요.
작성하신 `handleUpdate` 메소드에는 id, Task, HTTP Response에 대한 정보가 모두 담겨져 있어요.

Task를 업데이트 하는 일이 주요 목적이니, 우선은 Task에만 집중해 봅시다.

그러면 Service 클래스에 아래처럼 코드를 작성할 수 있을 거예요.
```
public Task handleUpdate(Task source) {
        Task task = findTask(id);
        task.setTitle(source.getTitle(id));

        return task;

    }
```

JIY0UNG committed:
TaskController 메서드들의 이름을 기능을 잘 표현할 수 있는 단어로 변경하라

- 기존에는 Task 수정 요청을 처리하는 두 메서드에 대해 '~ByPutMethod'와 같은 이름을 붙여주어 구분하도록 했었으나, 더 명료하면서도 그 기능을 명확히 표현할 수 있는 이름이 더 좋기 때문에 'update', 'patch'로 변경하였습니다.
- 이 외의 메서드들도 그들이 하는 '일'을 더 잘 나타낼 수 있는 이름들로 변경하였습니다.

JIY0UNG committed:
컨트롤러의 작업을 서비스에게 위임하라

- 기존에는 컨트롤러가 서비스 로직을 모두 포함하고 있었는데, 이는 응집도가 낮아진다는 문제점이 있습니다. 그래서 컨트롤러가 하던 일들을 서비스 객체로 분리하였습니다.

JIY0UNG committed:
클라이언트 측에 응답하는 코드를 분리하라

생성, 조회, 수정, 삭제 외의 일을 서비스 메서드에서 처리하는 것은 객체 지향적이지 않다고 생각했습니다.
따라서 서비스 메서드는 조건에 따라 적절한 응답 객체를 호출하여 content만 넘겨주고, 각각의 응답 객체들이 그에 따른 응답 코드를 설정 후 content를 반환하는 작업을 하도록 하였습니다.

JIY0UNG committed:
할 일을 추가하고 삭제할 수 있는 웹 페이지를 만들어라

JIY0UNG commented:
안녕하세요!
저번에 해주신 리뷰에 따라 메서드명을 변경했고, 기존에 컨트롤러가 하던 일들을 서비스 객체에 위임하도록 했습니다. 그리고 할 일을 추가하고 삭제할 수 있는 웹 페이지를 만드는 것까지 해봤어요!

객체의 역할과 책임에 유의하며 코드를 수정해 봤는데, 맞게 한 건지 자신이 없네요😥
리뷰 부탁드립니다~~

JIY0UNG committed:
컨트롤러의 getList 메서드의 반환형을 List<Task>로 바꿔라

다른 메서드들은 조건에 따라 Task 객체를 반환할 수도, String을 반환할 수도 있기에 반환형을 Object로 설정했습니다. 반면 getList 메서드는 항상 List<Task>만 반환하기 때문에 반환형이 Object일 필요가 없다고 생각하여 수정했습니다.

bloomspes commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -45,19 +45,19 @@ public Object createTask(@RequestBody Task task, HttpServletResponse response) {
-    public Object updateTaskByPutMethod(@PathVariable("id") Long id,
-                                        @RequestBody Task source, HttpServletResponse response) {
+    public Object update(@PathVariable("id") Long id,
+                         @RequestBody Task source, HttpServletResponse response) {
```
잘 변경하셨어요, 나머지 메소드도 모두 깔끔하네요. 👏🏻

bloomspes replied:
LGTM

JIY0UNG committed:
삭제 버튼을 누를 때 해당 Task의 id를 넘기는 방식을 수정하라

기존에는 hidden type으로 id를 넘겨줬었는데, 실행했을 때 제가 삭제하려고 했던 id와 다른 숫자가 넘겨져서 다른 Task가 삭제되는 문제가 있었습니다.
그래서 버튼의 on-click() 속성에 넘겨줄 인자도 같이 써주는 방식으로 수정하였고, 이전의 문제를 해결할 수 있었습니다.

bloomspes commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -1,106 +1,48 @@
-        return tasks;
+    public Object getList() {
+        return taskService.handleList();
     }
```
모든 작업을 반환하려면 어떤 타입을 사용해야 할 까요?

bloomspes commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -16,7 +17,7 @@ public TaskController(TaskService taskService) {
 
     @GetMapping
-    public Object getList() {
+    public List<Task> getList() {
         return taskService.handleList();
     }
```
아주 좋습니다. 👍🏻 나머지 메소드에 대해서도 `Object` 타입 대신에 어떤 타입을 사용하면 좋을 지 함께 고민해보세요.

bloomspes reviewed (commented):
커밋 메시지에 달아주신 의도도 명확하게 적어주셔서 잘 파악이 되네요.
HTTP Response를 Controller에서 관심사를 분리하고 캡슐화 까지 잘 하셨습니다. 👍🏻

JIY0UNG committed:
웹 페이지에서 할 일을 수정할 수 있도록 해라

- '수정' 버튼을 누르고 입력창에서 제목을 수정한 후, '확인' 버튼을 누르면 해당 Task의 id와 새로 입력한 제목을 가지고 patch 요청을 보낼 수 있도록 했습니다.
- 수정 중 '취소' 버튼을 누르면 다시 이전 페이지로 돌아가도록 했습니다.

bloomspes commented:
과제를 진행하는 데 어려움은 없으신가요? 오늘도 화이팅입니다. 👍🏻

JIY0UNG committed:
컨트롤러의 생성, 수정, 삭제 메서드의 반환값을 변경하라

- 기존에는 Task를 찾으면 Task 객체를 반환하고, 찾지 못했으면 빈 문자열을 반환해야 해서 메서드의 반환형을 Object로 통일했었습니다. 그러나 메서드가 반환하는 것이 무엇인지 명확히 알 수 없다는 문제점이 있었습니다.
그래서 생성, 수정 메서드에서 Task 객체를 반환하되, Task 객체가 null이면 그에 대한 예외를 발생시키도록 Optional 클래스로 감싸주었습니다.
- 삭제 메서드는 반환형을 boolean으로 수정하여, 삭제 여부를 반환하도록 하였습니다.
- 기존에 HttpServletResponse를 매개변수로 받아서 응답 코드를 직접 넣어줬던 것은 @ResponseStatus 어노테이션으로 수정하였습니다.

JIY0UNG commented:
안녕하세요~~
기존에 컨트롤러에서 Object를 반환하던 메서드들의 반환형을 Optional로 변경하고,
잘못된 요청에 대해 Response 객체들을 반환하도록 했던 것을 예외를 발생시키도록 변경했습니다!
삭제 메서드는 지금 제대로 작동하지 않는 것 같아서 이따가 다시 수정해야 할 것 같아요.
리뷰 부탁드립니다! 좋은 주말 보내세요😄

bloomspes commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -22,29 +24,31 @@ public List<Task> getList() {
     @GetMapping("{id}")
-    public Object getItem(@PathVariable("id") Long id, HttpServletResponse response) {
-        return taskService.handleItem(id, response);
+    public Optional<Task> getItem(@PathVariable("id") Long id) {
+        return taskService.handleItem(id);
```
`Optional<T>` 를 사용해서 값이 null 인 경우, `new NullPointException()` 객체를 반환하도록 의도를 잘 드러내셨어요. 👍🏻

bloomspes commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,70 @@
+    private List<Task> tasks = new ArrayList<>();
+    private Long newId = 0L;
+
+    public List<Task> handleList() {
+        return tasks;
+    }
```
모든 `Task`를 `List<T>`로 가져오는 메소드 이므로 이름을 `getTaskList()` 혹은 `getList()` 로 변경하면 의도를 드러낼 수 있을 것 같아요.

JIY0UNG replied:
앗 그렇네요! 지금 보니까 다른 메서드들도 좀 더 좋은 이름으로 수정해야겠어요

bloomspes commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -0,0 +1,70 @@
+        return task;
+    }
+
+    private void checkTitle(String title) {
+        if(title.isBlank()) {
+            throw new BadRequestException();
+        }
+    }
```
`isBlank()` 메소드를 사용하셨네요! 👍🏻 문자열의 길이를 받을 때 0이 아니라 비어있는 경우, 공백으로 채워져 있는 경우도 고려하는 건 참 좋습니다.

`TaskService` 클래스는 직접 Task가 생성, 조회, 수정, 삭제가 되어서 데이터베이스에 변경이 일어나는 곳 이예요.
`title`의 길이가 0인지, 값이 없는지를 체크하는 메소드는 어디에 위치하는 게 좋을까요?

JIY0UNG replied:
> title의 길이가 0인지, 값이 없는지를 체크하는 메소드는 어디에 위치하는 게 좋을까요?

컨트롤러 클래스는 비즈니스 로직을 직접 다루지 않고, 서비스 클래스에서는 DB 변경 작업이 직접 이루어지는 곳이니까, Task 클래스에 위치해야 하는 걸까요..?

JIY0UNG committed:
TaskService 클래스의 addTask 메서드를 제거하라

Task 생성 작업을 addTask 메서드로 굳이 나눌 필요가 없는 것 같다고 생각되어서, 기존에 addTask에서 하던 작업을 handleCreate 내부에서 하도록 수정했습니다.

JIY0UNG committed:
서비스 클래스의 리스트 조회, 아이템 조회, 생성, 수정, 삭제 메서드명을 변경하라

기존에 메서드 이름 앞에 handle을 붙였었는데, 각 메서드의 역할을 잘 드러낼 수 있는 이름으로 변경했습니다.

JIY0UNG committed:
서비스 클래스의 삭제 메서드에서 삭제할 Task를 get으로 가져와라

기존에는 Optional로 감싸진 Task 객체를 삭제하게 되어있었기 때문에, Tasks 리스트에서 해당 객체를 찾지 못해 삭제가 제대로 이루어지지 않고 있었습니다.
그래서 get()으로 Task 객체를 가져오도록 수정하였습니다.

JIY0UNG committed:
예외 클래스의 불필요한 주석들을 제거하라

예외 클래스의 메서드 중 다른 곳에서 호출되지 않아 주석 처리 해놨던 불필요한 메서드들을 삭제했습니다.

JIY0UNG committed:
Optional<Task> 형태로 반환하고 있는 메서드들의 반환형을 Task로 변경하라

Optional이 벗겨진 Task 객체를 얻기 위해서는 get()을 사용하면 된다는 것을 깨달았습니다. 그래서 아이템 조회, 수정 메서드가 Task 객체를 반환하도록 변경했습니다.

JIY0UNG committed:
서비스 클래스의 checkTItle메서드를 Task 클래스로 옮겨라

기존에는 입력한 title값이 비어있거나 빈 공백으로 이루어져 있는지 판별하는 checkTitle 메서드를 서비스 클래스에 위치시켰었는데, 직접적으로 DB를 다루는 서비스 클래스보다는 Task 클래스에 위치시키는 것이 좋다는 생각이 들어서 옮기게 되었습니다.

bloomspes commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -20,35 +20,35 @@ public TaskController(TaskService taskService) {
     @GetMapping
     public List<Task> getList() {
-        return taskService.handleList();
+        return taskService.getTaskList();
```
잘 변경 하셨어요. 👍🏻

bloomspes commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -20,35 +20,35 @@ public TaskController(TaskService taskService) {
-    public Optional<Task> getItem(@PathVariable("id") Long id) {
-        return taskService.handleItem(id);
+    public Optional<Task> getTaskItem(@PathVariable("id") Long id) {
+        return taskService.getTaskItem(id);
```
`TaskItem`은 무슨 뜻 일까요? 우리는 태스크를 만들면서 별도로 Item 을 분리하지는 않았을거예요, 이런 경우에는 나중에 설계하시다가 의미가 혼동이 생길 수가 있어요. task를 가져오고 싶다면 `getTask()`를 사용해 보세요.

JIY0UNG replied:
> 나중에 설계하시다가 의미가 혼동이 생길 수가 있어요

오 그 부분까지는 미처 생각하지 못했네요. 수정하겠습니다!

bloomspes commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -31,20 +31,21 @@ public Task addTask(Task task) {
         task.ifPresent(t -> t.setTitle(source.getTitle()));
 
-        return task;
+        return task.get();
```
반환을 `task`가 아닌, `task.get()`으로 가져오는 이유가 있을까요? 👀

JIY0UNG replied:
기존의 findTask메서드에서 Task를 찾지 못한 경우(null인 경우), NullPointerException을 던지지 않고 그냥 비어있는 객체를 가져오기 위해 `ofNullable()`을 사용했었습니다.
그런데 `ofNullable()`은 Optional 객체를 반환하기 때문에, findTask메서드의 반환형은 ` Optional<Task>`였어요.

따라서 findTask 메서드로 가져온 객체를 `Optional<Task>`가 아닌 Task의 형태로 반환하고 싶으면, `get()`으로 Optional이 담고 있는 Task 객체를 꺼내야 했습니다.

지금은 아래 코드처럼 `ofNullable()`을 쓸 필요 없이,
비어있는 객체에 대해서 `orElseThrow()`로 예외 처리를 해주면
`Optional<Task>`가 아닌 Task를 반환할 수 있다는 것을 깨달았습니다.
```java
public Task getTask(Long id) {
    return tasks.stream()
            .filter(task -> task.getId().equals(id))
            .findFirst()
            .orElseThrow(() -> new NotFoundException(id));
}
```
그래서 기존에 `task.get()`으로 반환하던 서비스 메서드들이 다시 `task`를 반환하도록 수정하고 있습니다!

bloomspes commented on app/src/main/java/com/codesoom/assignment/services/TaskService.java:
```diff
@@ -31,20 +31,21 @@ public Task addTask(Task task) {
     }
 
     public boolean deleteTask(Long id) {
         if(id == null) {
             throw new NotFoundException();
         }
         Optional<Task> task = findTask(id);
+
         return tasks.remove(task.get());
     }
```
우리는 id에 대응하는 `Task` 객체를 찾을 수 있고, `Tasks`에서 이 객체를 삭제할 수 있어요.
그러면 `Optional<Task>` 자료형을 사용하지 않고도 삭제가 가능할 거예요.

bloomspes replied:
```
Task task = getTask(id);
tasks.remove(task);
```
훨씬 간단하지요?

JIY0UNG committed:
task.get()을 반환하던 서비스 메서드들이 task를 반환하도록 변경해라

기존의 findTask 메서드에서 Task를 찾지 못한 경우 NPE를 던지는 대신 그냥 비어있는 객체를 반환하도록 하기 위해 ofNullable()을 사용했었습니다.
그런데 ofNullable()은 Optional 객체를 반환하기 때문에, findTask 메서드로 가져온 객체를 Task의 형태로 반환하려면 get()을 써야 했습니다.

get()을 쓰지 않고도 Task 객체를 반환할 수 있도록 하기 위해서,
ofNullable()을 사용하지 않고 orElseThrow()로 예외 처리를 해주어 findTask 메서드가 Task를 반환하도록 수정했습니다.

JIY0UNG committed:
ControllerAdvice로 예외 처리를 하라

컨트롤러에서 발생한 예외들을 한곳에서 관리할 수 있도록 ControllerAdvice를 사용했습니다.

JIY0UNG committed:
getTaskItem이라는 이름을 getTask로 변경하라

- 'Task Item'이라는 것이 따로 존재한다는 오해가 생길 수 있으므로 이름을 변경하였습니다.
- 기존에는 서비스 메서드에서 특정 Task 조회 작업을 하는 메서드를 findTask로 분리했었는데, 굳이 분리할 필요가 없다고 생각되어 getTask메서드 하나로 합쳤습니다.

JIY0UNG committed:
프론트 페이지 코드를 수정하라

기존 코드와 비슷하게 작동하지만, 풀이 강의에서 아샬님께서 보여주신 코드가 더 좋은 것 같아서 참고하며 수정했습니다.
